plugin/clouddns/clouddns.go

```diff
@@ -0,0 +1,275 @@
+			// We want the *longest* matching zone, otherwise we may end up in a parent
+			if len(zname) > len(zone) {
+				zone = zname
+
+			}
+		}
+	}
+	log.Infof("Returned zone name is %v", zone)
+
+	return zone
+}
+
+// ServeDNS implements the plugin.Handler.ServeDNS.
+func (h *CloudDNS) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+	log.Infof("request comparing is %v", qname)
+	for _, zo := range h.zoneNames {
+		log.Infof("zoneNames part is %v", zo)
+	}
+	zName := Zones(h.zoneNames).Matches(qname)
+	log.Infof("request compare result = (%v)", zName)
+
+	if zName == "" {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+	z, ok := h.zones[zName]
+	if !ok || z == nil {
+		log.Infof("Failed to find it ! (%v)", z)
+
+		return dns.RcodeServerFailure, nil
+	}
+
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative, m.RecursionAvailable = true, true
+	var result file.Result
+	for _, managedZone := range z {
+		h.zMu.RLock()
+		m.Answer, m.Ns, m.Extra, result = managedZone.z.Lookup(state, qname)
+		h.zMu.RUnlock()
+		if len(m.Answer) != 0 {
+			break
+		}
+	}
+
+	if len(m.Answer) == 0 && h.Fall.Through(qname) {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+
+	switch result {
+	case file.Success:
+	case file.NoData:
+	case file.NameError:
+		m.Rcode = dns.RcodeNameError
+	case file.Delegation:
+		m.Authoritative = false
+	case file.ServerFailure:
+		return dns.RcodeServerFailure, nil
+	}
+
+	w.WriteMsg(m)
+	return dns.RcodeSuccess, nil
+}
+
+// Run executes first update, spins up an update forever-loop.
+// Returns error if first update fails.
+func (h *CloudDNS) Run(ctx context.Context) error {
+	log.Infof("Entering Run function")
+	log.Info(h.zones)
+	if err := h.updateZones(ctx); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			log.Infof("Entering Run function loop")
+			log.Info(h.zones)
+			for i, j := range h.zones {
+				log.Infof("Zone index is", i)
+				log.Infof("Zone value is", j)
+				for k, l := range j {
+					log.Infof("Zone subindex is", k)
+					log.Infof("Zone dns is", l.dns)
+					log.Infof("Zone id is", l.id)
+					log.Infof("Zone file is", l.z.All())
+
+				}
+
+			}
+			log.Infof("Zones are %v", h.zones)
+			log.Infof("Zones length is %v", len(h.zones))
+
+			log.Infof("Zone names are %v", h.zoneNames)
+			log.Infof("Zones name length is %v", len(h.zoneNames))
+			select {
+			case <-ctx.Done():
+				log.Infof("Breaking out of CloudDNS update loop: %v", ctx.Err())
+				return
+			case <-time.After(1 * time.Minute):
+				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil /* Don't log error if ctx expired. */ {
+					log.Errorf("Failed to update zones: %v", err)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (h *CloudDNS) updateZones(ctx context.Context) error {
+	errc := make(chan error)
+	defer close(errc)
+	for zName, z := range h.zones {
+		go func(zName string, z []*zone) {
```

Comment:
Please add the following code above `go func`:

```
zName := zName
z := z
```

The reason is described at https://golang.org/doc/faq#closures_and_goroutines.

Reply:
Thanks!